### DISCUSSION TITLE
reorder events

### DISCUSSION THREAD

**Discussion by bmitzkus** [micromark/micromark#78] [Q&A] [answered]
Is there any way to reorder the events (like it is done for definitions in [compile](https://github.com/micromark/micromark/blob/30adcad68429a5c8e069f169f7735bc7bc018aee/packages/micromark/dev/lib/compile.js#L296)) in order to work with custom definitions?

**Comment by wooorm** [comment, marked as answer]
Sounds like somewhat an XY problem.
Footnotes also work: https://github.com/micromark/micromark-extension-footnote

**Reply by bmitzkus** [reply]
Sorry, let me elaborate: I want to provide custom syntax for reusable tooltips, much like URL/image rendering works. 

I took much inspiration from your footnote extension, however for footnotes rendering the reference before the footnote content works because it is just rendered as a reference id. In my case, I need something like the syntax for URLs/Images. For this to work, I would either need to use definition syntax as is, but access the mediaStack/definitions in `compile.js` from outside or have some way of reordering the events after tokenizing in order to have my custom definitions rendered before the references.

**Reply by wooorm** [reply]
I wrote a lot about why not to create syntax extensions: https://github.com/micromark/micromark#extending-markdown

Why not use directives?

**Reply by bmitzkus** [reply]
I took a look at directives, but I don't see how they solve my problem. Can I create reusable content with directives? In my use case, tooltips are used to explain terminology in articles, where the same term can occur often and I would like to avoid rewriting/copying the explanation over for each occurence, thats why I was looking for something like

```markdown
[url]

[url]: example.com
```

where `url` can also be reused in the text.

**Reply by wooorm** [reply]
Directives can do basically anything.

I’d suggest:

```
Lorem ipsum.:tip{#identifier} Or :tip[a tip around content]{#identifier}

:::description{#identifier}
# Description for thing

All the markdown content you want in here.
:::
```

**Reply by bmitzkus** [reply]
Oh wow, thanks a lot for the examples, I'll take a look at it again.

I'm working with the unified stack for only two weeks now and I am impressed by the flexibility and overall quality of all the projects. Thanks for putting so much work in it =)

**Reply by ChristianMurphy** [reply]
Reference links already offer that syntax today https://spec.commonmark.org/0.30/#reference-link
maybe use [`remark`](https://github.com/remarkjs/remark) (which builds on micromark) to render `linkReference` nodes with your custom tooltip? https://astexplorer.net/#/gist/d5d88539bc0bf6b146e31dc08a7a2551/2347487753c4573fc58846c3a16c688c7759e9f9

or use a pair of directives.

```markdown
:tooltip[something]

:tooltipDefinition[something]{content="description"}
```
